Module1.FileSystemVisitor/FileSystemVisitor/FileSystemVisitor.cs

```diff
@@ -0,0 +1,179 @@
+
+        public event EventHandler<WorkEventArgs> WorkStarted;
+
+        public event EventHandler<WorkEventArgs> WorkFinished;
+
+        public event EventHandler<FileFoundEventArgs> FileFound;
+
+        public event EventHandler<FileFoundEventArgs> DirectoryFound;
+
+        public event EventHandler<FileFoundEventArgs> FilteredFileFound;
+
+        public event EventHandler<FileFoundEventArgs> FilteredDirectoryFound;
+
+        #region Constructors
+
+        public FileSystemVisitor()
+        {
+            _func = (x) => true;
+        }
+
+        public FileSystemVisitor(Func<FileSystemInfo, bool> func)
+        {
+            _func = func;
+        }
+
+        #endregion
+
+        public void CompleteSearch()
+        {
+            _isSearchComplete = true;
+        }
+
+        public void SkipFile(string fileName)
+        {
+            _nextFileToSkip = fileName;
+        }
+
+        public IEnumerable<string> GetFileList(string directoryPath)
+        {
+            OnWorkStarted(this, new WorkEventArgs("Work started."));
+
+            var directoryInfo = new DirectoryInfo(directoryPath);
+            foreach (var file in directoryInfo.GetFiles())
+            {
+                OnFileFound(this, new FileFoundEventArgs("File found.", file.FullName));
+
+                if (_func(file))
+                {
+                    OnFilteredFileFound(this, new FileFoundEventArgs("Filtered file found.", file.FullName));
+
+                    if (_isSearchComplete)
+                    {
+                        yield break;
+                    }
+
+                    if (file.FullName.Equals(_nextFileToSkip))
+                    {
+                        continue;
+                    }
+
+                    yield return file.FullName;
+                }
+            }
+
+            foreach (var directory in directoryInfo.GetDirectories())
+            {
+                OnDirectoryFound(this, new FileFoundEventArgs("Directory found.", directory.FullName));
+                if (_func(directory))
+                {
+                    OnFilteredDirectoryFound(this, new FileFoundEventArgs("Directory found.", directory.FullName));
+
+                    if (_isSearchComplete)
+                    {
+                        yield break;
+                    }
+
+                    if (directory.FullName.Equals(_nextFileToSkip))
+                    {
+                        continue;
+                    }
+
+                    foreach(var file in GetFileList(directory.FullName))
+                    {
+                        yield return file;
+                    }
+
+                    yield return directory.FullName;
+                }
+            }
+
+            OnWorkFinished(this, new WorkEventArgs("Work finished."));
+        }
+
+        protected virtual void OnWorkStarted(object sender, WorkEventArgs e)
+        {
+            if(WorkStarted != null)
+            {
+                WorkStarted(this, e);
```

Comment:
WorkStarted?.Invoke(this, e) shortly and look better in my opinion

Reply:
done